src/input_common/gcadapter/gc_poller.cpp

```diff
@@ -0,0 +1,301 @@
+                params.Set("button", GCAdapter::PAD_BUTTON_UP);
+                break;
+            }
+            if (pad.button & GCAdapter::PAD_TRIGGER_L) {
+                params.Set("button", GCAdapter::PAD_TRIGGER_L);
+                break;
+            }
+            if (pad.button & GCAdapter::PAD_TRIGGER_R) {
+                params.Set("button", GCAdapter::PAD_TRIGGER_R);
+                break;
+            }
+            if (pad.button & GCAdapter::PAD_TRIGGER_Z) {
+                params.Set("button", GCAdapter::PAD_TRIGGER_Z);
+                break;
+            }
+            if (pad.button & GCAdapter::PAD_BUTTON_START) {
+                params.Set("button", GCAdapter::PAD_BUTTON_START);
+                break;
+            }
+            // For Axis button implementation
+            if (pad.axis != GCAdapter::PadAxes::Undefined) {
+                params.Set("axis", static_cast<u8>(pad.axis));
+                params.Set("button", GCAdapter::PAD_STICK);
+                if (pad.axis_value > 128) {
+                    params.Set("direction", "+");
+                    params.Set("threshold", "0.5");
+                } else {
+                    params.Set("direction", "-");
+                    params.Set("threshold", "-0.5");
+                }
+                break;
+            }
+        }
+    }
+    return params;
+}
+
+void GCButtonFactory::BeginConfiguration() {
+    polling = true;
+    adapter->BeginConfiguration();
+}
+
+void GCButtonFactory::EndConfiguration() {
+    polling = false;
+    adapter->EndConfiguration();
+}
+
+class GCAnalog final : public Input::AnalogDevice {
+public:
+    GCAnalog(int port_, int axis_x_, int axis_y_, float deadzone_,
+             std::shared_ptr<GCAdapter::Adapter> adapter)
+        : port(port_), axis_x(axis_x_), axis_y(axis_y_), deadzone(deadzone_), gcadapter(adapter) {}
+
+    float GetAxis(int axis) const {
+        std::lock_guard lock{mutex};
+        // division is not by a perfect 128 to account for some variance in center location
+        // e.g. my device idled at 131 in X, 120 in Y, and full range of motion was in range
+        // [20-230]
+        return (gcadapter->GetPadState()[port].axes.at(axis) - 128.0f) / 95.0f;
+    }
+
+    std::tuple<float, float> GetAnalog(int axis_x, int axis_y) const {
+        float x = GetAxis(axis_x);
+        float y = GetAxis(axis_y);
+
+        // Make sure the coordinates are in the unit circle,
+        // otherwise normalize it.
+        float r = x * x + y * y;
+        if (r > 1.0f) {
+            r = std::sqrt(r);
+            x /= r;
+            y /= r;
+        }
+
+        return std::make_tuple(x, y);
+    }
+
+    std::tuple<float, float> GetStatus() const override {
+        const auto [x, y] = GetAnalog(axis_x, axis_y);
+        const float r = std::sqrt((x * x) + (y * y));
+        if (r > deadzone) {
+            return std::make_tuple(x / r * (r - deadzone) / (1 - deadzone),
```

Comment:
This can just use bracing syntax

```suggestion
            return {x / r * (r - deadzone) / (1 - deadzone),
```